Add communication data attributes to Zwave node_entity

turbokongen opened the pull request:
## Description:
This adds communication data attributes to the zwave.node_entity
This is as follows:
```
         sentCnt                             # Number of messages sent from this node.
         sentFailed                          # Number of sent messages failed
         retries                            # Number of message retries
         receivedCnt                       # Number of messages received from this node.
         receivedDups                      # Number of duplicated messages received;
         receivedUnsolicited              # Number of messages received unsolicited
         sentTS                                 # Last message sent time
         receivedTS                            # Last message received time
         lastRequestRTT                  # Last message request RTT
         averageRequestRTT             # Average Request Round Trip Time (ms).
         lastResponseRTT                 # Last message response RTT
         averageResponseRTT           #Average Reponse round trip time.
```

turbokongen committed:
Add quality attribute to node

turbokongen committed:
Move quality to node_entity

turbokongen committed:
adjustments

turbokongen committed:
Line lenght

turbokongen committed:
flake8

andrey-git commented on homeassistant/components/zwave/node_entity.py:
```diff
@@ -95,6 +98,33 @@ def network_node_changed(self, node=None, args=None):
+        maxrtt = 10000.0
+        data = self.get_node_statistics()
+        if data and data != {}:
+            data1 = (float(float(data['sentCnt'] - data['sentFailed']) /
```
Why convert to float twice?

andrey-git commented on homeassistant/components/zwave/node_entity.py:
```diff
@@ -95,6 +98,33 @@ def network_node_changed(self, node=None, args=None):
+                                      data['receivedUnsolicited']) /
+                                data['receivedCnt'])) if data['receivedCnt']
+                     != 0 else 0.0)
+            quality = ((data1 + (data2*2) + (data3*3) + data4) / 7.0) * 100.0
```
Why this magic formula?

andrey-git reviewed (commented):
I'm all for exposing more data, but why this magic formula?

The transmitted/failed rate is useful on its own. The RTT time is also useful on its own.
Another useful thing (not sure if it fits as a node property) is connectivity: direct to controller=best.
1 hop via multiple nodes worse, 2 hops via multiple or 1 hop via single worse, etc.

Also seems GetNodeStatistics already have a `quality` field. What's in that field?

turbokongen commented:
The quality field is always 0

turbokongen commented:
Or should I maybe just replace quality attribute with the complete output of the statistics as one attribute?

andrey-git commented:
Maybe take GetNodeStatistics output and clean it up?
I.e. remove fields that are always zero, empty, etc

@armills what do you think?

armills commented:
I'm with @andrey-git here. The underlying information can be exposed directly as their own attributes, and this percentage could always be recreated as a template sensor. We could possibly put an example in the docs. I think a percentage would be more useful in a signal strength type situation, where we're just trying to normalize a reading on a different scale.

Also, I'm not sure if we have a good precedence on where we draw the line as far as exposing information to hass. Is anyone going to automate based on these attributes? It seems to me that we already expose `is_failed`, and at that point it could be possible for users to examine the ozw log to see what was happening. I think once we decide on a direction, we should codify it somewhere in the development docs to make sure everyone is working towards the same goal.

andrey-git commented:
Note that `is_failed` is the opinion of the controller which is not very precise.
I have a lock node that is disconnected for weeks and the controller still thinks it is OK.
I have a dimmer with a flaky connectivity and the controller often marks it as dead.

turbokongen commented:
The output of getNodeStatistics are:

```
         sentCnt                             # Number of messages sent from this node.
         sentFailed                          # Number of sent messages failed
         retries                            # Number of message retries
         receivedCnt                       # Number of messages received from this node.
         receivedDups                      # Number of duplicated messages received;
         receivedUnsolicited              # Number of messages received unsolicited
         sentTS                                 # Last message sent time
         receivedTS                            # Last message received time
         lastRequestRTT                  # Last message request RTT
         averageRequestRTT             # Average Request Round Trip Time (ms).
         lastResponseRTT                 # Last message response RTT
         averageResponseRTT           #Average Reponse round trip time.
         quality                                # Node quality measure, always stays at 0
         lastReceivedMessage[254]      # Place to hold last received message
         ccData                               # List of statistic on each command_class
            commandClassId   # Num type of CommandClass id.
            sentCnt             # Number of messages sent from this CommandClass.
            receivedCnt        # Number of messages received from this CommandClass.
```

andrey-git commented:
Are all those actually filled? If yes, lets keep them all except lastReceivedMessage and ccData as I expect those to be too large.

turbokongen commented:
Here's an example:
```
communication data
{ "receivedCnt": 3, "ccData": [ { "receivedCnt": 0, "commandClassId": 134, "sentCnt": 0 }, { "receivedCnt": 1, "commandClassId": 133, "sentCnt": 1 }, { "receivedCnt": 1, "commandClassId": 115, "sentCnt": 1 }, { "receivedCnt": 0, "commandClassId": 114, "sentCnt": 0 }, { "receivedCnt": 0, "commandClassId": 112, "sentCnt": 0 }, { "receivedCnt": 1, "commandClassId": 32, "sentCnt": 1 }, { "receivedCnt": 0, "commandClassId": 0, "sentCnt": 0 } ], "receivedUnsolicited": 0, "sentTS": "2017-03-27 15:38:15:620 ", "averageRequestRTT": 2462, "lastResponseRTT": 3679, "retries": 0, "sentFailed": 1, "sentCnt": 7, "quality": 0, "lastRequestRTT": 1591, "lastReceivedMessage": [ 0, 4, 0, 15, 3, 32, 3, 0, 221, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 ], "receivedDups": 0, "averageResponseRTT": 2443, "receivedTS": "2017-03-27 15:38:19:298 " }
```

andrey-git commented:
The problem with exposing this data as attribute is that it will kill the recorder, as the attributes will change very very often.

andrey-git commented:
Actually, not *that* often.

So should each field like `sentCnt` be an attribute, or a single dict attribute with everything?

turbokongen commented:
So, how about we save this for when we make a zwave control panel?

andrey-git commented:
ok. 

I'm thinking about automation or a built-in that would refresh / heal the network when a device is "lost"...

turbokongen committed:
Cleanup and add each entry from getNodeStatistics as attribute

turbokongen commented:
Cleaned up and made it like this:
![image](https://cloud.githubusercontent.com/assets/16622752/24376645/e2ecd536-133c-11e7-8f4a-351d694d5b08.png)
Does that make sense?

andrey-git commented:
Seems good.

Could you let it run for a while to see if this generates too much event spam?

We could make a list in the config at zwave level (not per device) to allow the user to specify a list of attributes they are interested in...

turbokongen commented:
It does not seem to generate any big deal of extra logging.
There is a state_changed for every message sent or recieved to the node.
My least active node(remotec_zxt120 ac_ir_controller) has 16 state changes since last night, and my very active node(aetoec 2nd edition micro switch with energy measuring) has 296 state_changed.
You could always exclude the state changes from the zwave domain to the recorder.

I need to update the tests, but can I just make a new class that will contain the data of getNodeStatistics for that?

turbokongen commented:
Eh, different problem: How to mock `NETWORK` ?

andrey-git commented:
If you use mock_openzwave param (like other tests), `NETWORK` should already be a MagicMock

turbokongen commented:
Okay... Small steps working forward.I must admit I don't quite understand how the Mock and MagicMock works, but I made that part work..  I'm not sure how to feed the values in the test though. I'm at here now:
```
E       AssertionError: {'is_info_received': True, 'query_stage': '[174 chars]alse} != {'battery_level': 42, 'query_stage': 'Dynam[1424 chars]56'>}
E       + {'averageRequestRTT': <MagicMock name='mock.manager.getNodeStatistics().__getitem__()' id='140621794964256'>,
E       +  'averageResponseRTT': <MagicMock name='mock.manager.getNodeStatistics().__getitem__()' id='140621794964256'>,
E       - {'battery_level': 42,
E       ? ^
E       
E       +  'battery_level': 42,
E       ? ^
E       
E          'is_awake': True,
E          'is_failed': False,
E          'is_info_received': True,
E          'is_ready': False,
E          'is_zwave_plus': False,
E       +  'lastRequestRTT': <MagicMock name='mock.manager.getNodeStatistics().__getitem__()' id='140621794964256'>,
E       +  'lastResponseRTT': <MagicMock name='mock.manager.getNodeStatistics().__getitem__()' id='140621794964256'>,
E          'max_baud_rate': 40000,
E          'node_id': '567',
E          'query_stage': 'Dynamic',
E       +  'receivedCnt': <MagicMock name='mock.manager.getNodeStatistics().__getitem__()' id='140621794964256'>,
E       +  'receivedDups': <MagicMock name='mock.manager.getNodeStatistics().__getitem__()' id='140621794964256'>,
E       +  'receivedTS': <MagicMock name='mock.manager.getNodeStatistics().__getitem__()' id='140621794964256'>,
E       +  'receivedUnsolicited': <MagicMock name='mock.manager.getNodeStatistics().__getitem__()' id='140621794964256'>,
E       +  'retries': <MagicMock name='mock.manager.getNodeStatistics().__getitem__()' id='140621794964256'>,
E       +  'sentCnt': <MagicMock name='mock.manager.getNodeStatistics().__getitem__()' id='140621794964256'>,
E       +  'sentFailed': <MagicMock name='mock.manager.getNodeStatistics().__getitem__()' id='140621794964256'>,
E       +  'sentTS': <MagicMock name='mock.manager.getNodeStatistics().__getitem__()' id='140621794964256'>,
E          'wake_up_interval': 1800}

tests/components/zwave/test_node_entity.py:116: AssertionError
 tests/components/zwave/test_node_entity.py âœ“âœ“âœ“âœ“âœ“âœ“âœ“âœ“â¨¯âœ“âœ“âœ“âœ“âœ“                                                                                                         100% â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ

Results (0.33s):
      13 passed
       1 failed

```

balloob commented on homeassistant/components/zwave/__init__.py:
```diff
@@ -295,7 +295,7 @@ def value_added(node, value):
     def node_added(node):
         """Called when a node is added on the network."""
-        entity = ZWaveNodeEntity(node)
+        entity = ZWaveNodeEntity(node, NETWORK)
```
We should not pass the network around. Preferably we migrate the `NETWORK` global to `hass.data[ZWAVE_NETWORK]`

balloob commented:
So how spammy will these attributes be? Should we debounce updates to them ?

andrey-git commented:
@balloob 
Note that the "regular" 100ms debounce applies here too.

What do you think about opt-in approach, so the user specifies (globally, not per device) which attributes they are interested in?

balloob commented:
I don't know what the best approach here is.

I do know however that this is one of those PRs that will instigate a PR to add more filters to the recorder ;-)

turbokongen committed:
Move NETWORK to hass.data

turbokongen committed:
Test Done

andrey-git commented on homeassistant/components/zwave/node_entity.py:
```diff
@@ -80,6 +85,7 @@ def __init__(self, node):
         self.wakeup_interval = None
         self.location = None
         self.battery_level = None
+        self.quality = None
```
No longer needed

andrey-git commented:
Looks good. Please make sure tests pass

balloob commented on homeassistant/components/zwave/__init__.py:
```diff
@@ -195,7 +194,7 @@ def get_config_value(node, value_index, tries=5):
 def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
     """Generic Z-Wave platform setup."""
-    if discovery_info is None or NETWORK is None:
+    if discovery_info is None or hass.data[ZWAVE_NETWORK] is None:
```
Should be `ZWAVE_NETWORK not in hass.data`

Otherwise you can get a key not found exception.

balloob replied:
Actually, this cannot happen because zwave is a dependency and it will only set up a platform if the main component set up correctly.

balloob commented on homeassistant/components/zwave/__init__.py:
```diff
@@ -251,7 +247,7 @@ def setup(hass, config):
     options.lock()
 
-    NETWORK = ZWaveNetwork(options, autostart=False)
+    hass.data[ZWAVE_NETWORK] = ZWaveNetwork(options, autostart=False)
```
You can keep all code below easier to read (and slightly faster due to skipping unnecessary dict lookup) if you do:

```python
network = hass.data[ZWAVE_NETWORK] = ZWaveNetwork(…)
```

turbokongen committed:
Cleanup from review

armills commented on tests/components/cover/test_zwave.py:
```diff
@@ -8,58 +8,61 @@
     values = MockEntityValues(primary=value, node=node)
 
-    device = zwave.get_device(node=node, values=values, node_config={})
+    device = zwave.get_device(hass, node=node, values=values, node_config={})
```
Let's use keyworded arguments here to match the invocation we're using in `zwave/__init__.py` https://github.com/turbokongen/home-assistant/blob/59dcb12a77da73f815f4e7b513d5123606b85310/homeassistant/components/zwave/__init__.py#L726

turbokongen committed:
Resolve conflicts

turbokongen committed:
Review changes

turbokongen committed:
Long lines....

turbokongen committed:
blank line

turbokongen committed:
import-error-disable

armills commented on homeassistant/components/zwave/__init__.py:
```diff
@@ -230,6 +226,7 @@ def setup(hass, config):
                       "https://home-assistant.io/components/zwave/")
         return False
     from pydispatch import dispatcher
+    # pylint: disable=import-error
```
Any idea why this is causing an error now, and not before?

turbokongen commented:
It looks like this was the only place it wasn't disabled. But why it
didn't present an error before, is beyond me.

30. mar. 2017 21:08 skrev "Adam Mills" <notifications@github.com>:

> *@armills* commented on this pull request.
> ------------------------------
>
> In homeassistant/components/zwave/__init__.py
> <https://github.com/home-assistant/home-assistant/pull/6813#discussion_r109011360>
> :
>
> > @@ -230,6 +226,7 @@ def setup(hass, config):
>                        "https://home-assistant.io/components/zwave/")
>          return False
>      from pydispatch import dispatcher
> +    # pylint: disable=import-error
>
> Any idea why this is causing an error now, and not before?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/6813#pullrequestreview-30085121>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AP2koAwCI6XcHU-rsp_BeVlQ08jfuAySks5rq_2JgaJpZM4MqPFH>
> .
>

armills commented:
Yeah, I took back the comment after posting. The disable was actually just moved from here: https://github.com/home-assistant/home-assistant/pull/6813/files#diff-34736f9c4da86b6334cf5d201370707fL219

balloob commented:
Before we go ahead and merge this, here is an idea I just had.

How often is this data valuable ? Is this something you will base your automation on or is this only valuable from time to time to do diagnostics?

Because if it is the latter, we can keep it out of the state machine and instead create a Z-Wave panel in the frontend that talks to Z-Wave over a Z-Wave rest API.

andrey-git commented:
Maybe make this opt-in then? Whoever wants automation will add it, and for others it won't pollute the state machine.

balloob commented:
You know I don't like options ;-)

What would be a use case where you want to automate or track history of averageResponseRTT ?

andrey-git commented:
I would want to get a notification if a node drops dead or starts to get too high latency.

I admit I would prefer a way to do it once and not per-node.

When and if I set up grafana I would probably like to push averageResponseRTT there too just out of curiosity. (And maybe see that the graph is not interesting and remove it)

balloob commented:
In that case, would it make sense to have Z-Wave add a persistent notification when a node drops dead ?

andrey-git commented:
Not as a built-in option. In my setup zwave is too eager to declare a node dead.

turbokongen committed:
resolve conflict

turbokongen committed:
Update tests Part1... Again

turbokongen committed:
Hound

turbokongen committed:
Argh!

turbokongen committed:
Argh!

turbokongen committed:
YABBADABBADOOOOOOO!

turbokongen committed:
Enhance tests

turbokongen committed:
hound

turbokongen committed:
Resolve

turbokongen committed:
resolve tests...

turbokongen commented:
So, are we good to go on this, or should it be omitted from the state machine?
And in that case, how do I do that?
